### PR TITLE
MNTOR-3117 - E2E Test Additions

### PR DESCRIPTION
**Pull request by rafeerahman**
# References: 
Jira: MNTOR-3117, see "in review" tickets

Includes: MNTOR-3122, MNTOR-3123, MNTOR-3125, MNTOR-3132, MNTOR-3139, MNTOR-3215

# Screenshot (if applicable)

Not applicable.

# How to test

Test against this branch in GHA.
Passing: https://github.com/mozilla/blurts-server/actions/runs/9113316608

# Checklist (Definition of Done)
- [ ] Localization strings (if needed) have been added.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [ ] I've added or updated the relevant sections in readme and/or code comments
- [ ] I've added a unit test to test for potential regressions of this bug.
- [x] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [x] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [x] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: src/e2e/specs/landing.spec.ts [hunk 1]
```diff
@@ -158,6 +158,73 @@ test.describe(`${process.env.E2E_TEST_ENV} - Verify the Landing Page content`, (
     await expect(landingPage.privacyLink).toBeVisible();
     await expect(landingPage.githubLink).toBeVisible();
   });
+
+  test("Verify the 'Get data removal' button UI and functionality for both yearly and montly options", async ({
```
Comment (pdehaan) [inline src/e2e/specs/landing.spec.ts:162]:
```suggestion
  test("Verify the 'Get data removal' button UI and functionality for both yearly and monthly options", async ({
```

File: src/e2e/specs/landing.spec.ts [hunk 2]
```diff
@@ -158,6 +158,73 @@ test.describe(`${process.env.E2E_TEST_ENV} - Verify the Landing Page content`, (
...
+      description:
+        "https://testrail.stage.mozaws.net/index.php?/cases/view/2463504",
+    });
+
+    await landingPage.monitorHeroFormEmailInputField.fill("invalid");
+    await landingPage.monitorHeroFormInputSubmitButton.click();
+    // Stays on same page
+    await expect(landingPage.monitorHeroFormEmailInputField).toBeVisible();
+
+    const randomEmail = `${Date.now()}_tstact@restmail.net`;
```
Comment (pdehaan) [inline src/e2e/specs/landing.spec.ts:207]: Doubt it matters much for anything, but the other restmail accounts looked like:
```
`_${Date.now()}@restmail.net`;
```

File: src/e2e/specs/purchase.spec.ts [hunk 3]
```diff
@@ -5,61 +5,37 @@
...
+      description:
+        "https://testrail.stage.mozaws.net/index.php?/cases/view/2463564",
+    });
+
+    // this test runs through the welcome scan flow, increasing timeout to address it
+    test.slow();
+
+    setEnvVariables(process.env.E2E_TEST_ACCOUNT_EMAIL as string);
+
+    // speed up test by ignore non necessary requests
```
Comment (pdehaan) [inline src/e2e/specs/purchase.spec.ts:21]:
```suggestion
    // speed up test by ignoring non necessary requests
```

File: src/e2e/specs/dashboard.spec.ts [hunk 4]
```diff
@@ -268,37 +273,78 @@ test.describe(`${process.env.E2E_TEST_ENV} - Breaches Dashboard  - Payment`, ()
...
+  });
+});
+
+test.describe(`${process.env.E2E_TEST_ENV} - Breaches Dashboard - Breaches Scan, Continuous Protection, Data Profile Actions`, () => {
+  test.use({ storageState: { cookies: [], origins: [] } });
+
+  test.beforeEach(async ({ landingPage, page, authPage, welcomePage }) => {
+    test.slow();
+
+    await page.route(/(analytics)/, async (route) => {
```
Comment (pdehaan) [inline src/e2e/specs/dashboard.spec.ts:287]:
```suggestion
    // speed up test by ignoring non necessary requests
    await page.route(/(analytics)/, async (route) => {
```

File: src/e2e/specs/dashboard.spec.ts [hunk 5]
```diff
@@ -268,37 +273,78 @@ test.describe(`${process.env.E2E_TEST_ENV} - Breaches Dashboard  - Payment`, ()
...
-      `${process.env.E2E_TEST_ENV === "prod" ? "yearly" : "every ⁨2⁩ months"}`,
+    await dashboardPage.verifyPremiumUpsellModalOptions();
+  });
+});
+
+test.describe(`${process.env.E2E_TEST_ENV} - Breaches Dashboard - Breaches Scan, Continuous Protection, Data Profile Actions`, () => {
+  test.use({ storageState: { cookies: [], origins: [] } });
+
+  test.beforeEach(async ({ landingPage, page, authPage, welcomePage }) => {
+    test.slow();
```
Comment (pdehaan) [inline src/e2e/specs/dashboard.spec.ts:285]:
```suggestion
    // this test runs through the welcome scan flow, increasing timeout to address it
    test.slow();
```

Comment (pdehaan) [reply src/e2e/specs/dashboard.spec.ts:285]: conversely, we can set the `description` within the `.slow()` annotation so it shows up in reports. Per https://playwright.dev/docs/api/class-test#test-slow

```suggestion
    test.slow(true, "this test runs through the welcome scan flow, increasing timeout to address it");
```

File: src/e2e/specs/purchase.spec.ts [hunk 6]
```diff
@@ -5,61 +5,37 @@
...
+test.describe(`${process.env.E2E_TEST_ENV} - Breach Scan, Monitor Plus Purchase Flow Yearly`, () => {
+  test.beforeEach(async ({ page, authPage, landingPage, welcomePage }) => {
+    test.info().annotations.push({
+      type: "testrail id",
+      description:
+        "https://testrail.stage.mozaws.net/index.php?/cases/view/2463564",
+    });
+
+    // this test runs through the welcome scan flow, increasing timeout to address it
+    test.slow();
```
Comment (pdehaan) [inline src/e2e/specs/purchase.spec.ts:17]: See previous comment about totally optional variant of adding the description to the `test.slow()`, ie:

```suggestion
    test.slow(true, "this test runs through the welcome scan flow, increasing timeout to address it");
```

File: src/e2e/utils/helpers.ts [hunk 7]
```diff
@@ -150,3 +150,13 @@ export const checkAuthState = async (page: Page) => {
...
 };
+
+/**
+ * Strings may sometimes contain invisible unicode char's that make it hard to do assertions.
+ *
+ * @param text
+ */
+export function removeUnicodeChars(text: string): string {
+  // eslint-disable-next-line no-control-regex
+  return text.replace(/[^\x00-\x7F]/g, "");
```
Comment (pdehaan) [inline src/e2e/utils/helpers.ts:161]: 👍

File: src/e2e/specs/dashboard.spec.ts [hunk 8]
```diff
@@ -268,37 +273,78 @@ test.describe(`${process.env.E2E_TEST_ENV} - Breaches Dashboard  - Payment`, ()
...
+  });
+
+  test("Verify that the Premium upsell modal is displayed correctly - Continuous Protection, verify that the user can mark Data broker profiles as fixed", async ({
+    dashboardPage,
+  }) => {
+    test.info().annotations.push(
+      {
+        type: "testrail id #1",
+        description:
+          "(continuous protection step) \n https://testrail.stage.mozaws.net/index.php?/cases/view/2463623",
```
Comment (pdehaan) [inline src/e2e/specs/dashboard.spec.ts:309]: Q: why the newline?

File: src/e2e/pages/dashBoardPage.ts [hunk 9]
```diff
@@ -190,4 +225,39 @@ export class DashboardPage {
...
+    await expect(purchasePage.planDetails).toContainText("monthly");
+    await expect(purchasePage.subscriptionHeader).toBeVisible();
+
+    // Check yearly redirection
+    await this.page.goto(`${process.env.E2E_TEST_BASE_URL}`);
+    await this.subscribeButton.waitFor();
+    await this.subscribeButton.click();
+    await this.subscribeDialogSelectYearlyPlanLink.click();
+    await purchasePage.subscriptionHeader.waitFor();
+    // Text contains invisible characters which need to be hardcoded in an assertion
```
Comment (pdehaan) [inline src/e2e/pages/dashBoardPage.ts:254]: Q: What does this comment mean? I see we have the `removeUnicodeChars(text: string): string {...}` below, but we aren't using it here to remove any invisible chars.

File: src/e2e/specs/dashboard.spec.ts (hunk 8 above)
Comment (rafeerahman) [reply src/e2e/specs/dashboard.spec.ts]: Removed

File: src/e2e/pages/dashBoardPage.ts (hunk 9 above)
Comment (rafeerahman) [reply src/e2e/pages/dashBoardPage.ts]: Using the function now and removed the comment, https://github.com/mozilla/blurts-server/pull/4543/commits/890aec5177df521b278422223a7db4aced5ae7ab#diff-75ef62711cbcf8a457647f69a04069d3de68ce6e40887e0ac2373ee64ce2651bL253-L257

**Comment by rafeerahman** [comment]
Addressed the changes now @pdehaan , thanks for the review

File: src/e2e/pages/landingPage.ts [hunk 10]
```diff
@@ -71,13 +71,37 @@ export class LandingPage {
...
+  readonly freeMonitoringTooltipText: Locator;
+  readonly monitorPlusTooltipTrigger: Locator;
+  readonly monitorPlusTooltipText: Locator;
+  readonly closeTooltips: Locator;
 
   constructor(page: Page) {
     this.page = page;
+    this.freeMonitoringTooltipTrigger = page
+      .getByRole("gridcell", { name: "Manual removal Open tooltip" })
+      .getByLabel("Open tooltip");
```
Comment (pdehaan) [inline src/e2e/pages/landingPage.ts:88]: :heart:

File: src/e2e/pages/welcomeScanPage.ts [hunk 11]
```diff
@@ -67,4 +68,28 @@ export class WelcomePage {
     this.modalConfirmButton = page.getByRole("button", { name: "Confirm" });
     this.modalEditButton = page.getByRole("button", { name: "Edit" });
   }
+
+  async goThroughFirstScan() {
+    // confirm get started step elements
+    expect(await this.getStartedStep.count()).toEqual(3);
+    await expect(this.page.getByText("Get started")).toBeVisible();
+    await expect(this.page.getByText("Enter info")).toBeVisible();
+    await expect(this.page.getByText("Find exposures")).toBeVisible();
```
Comment (pdehaan) [inline src/e2e/pages/welcomeScanPage.ts:77]: Not sure if these locators should be in a POM, but we can also file a followup ticket to refactor in "the future" and land this for now.

File: src/e2e/specs/dashboard.spec.ts [hunk 12]
```diff
@@ -269,38 +274,83 @@ test.describe(`${process.env.E2E_TEST_ENV} - Breaches Dashboard  - Payment`, ()
...
+
+    await landingPage.open();
+    await landingPage.goToSignIn();
+
+    const randomEmail = `_${Date.now()}@restmail.net`;
+    await authPage.signUp(randomEmail, page);
+
+    await page.waitForURL("**/user/welcome");
+    await welcomePage.goThroughFirstScan();
+    expect(page.url()).toContain("/user/dashboard");
```
Comment (pdehaan) [inline src/e2e/specs/dashboard.spec.ts:304]: nit: we could also use `expect(page).toHaveURL(regexpOrString)`, but this works too.

File: src/e2e/specs/dashboard.spec.ts [hunk 13]
```diff
@@ -269,38 +274,83 @@ test.describe(`${process.env.E2E_TEST_ENV} - Breaches Dashboard  - Payment`, ()
...
+      }
+    }
+
+    await dashboardPage.skipExposureRemoval.click();
+    await dashboardPage.continuousProtectionButton.waitFor();
+    await expect(dashboardPage.continuousProtectionButton).toBeVisible();
+    await dashboardPage.continuousProtectionButton.click();
+    await dashboardPage.verifyPremiumUpsellModalOptions();
+
+    expect(await dashboardPage.numExposures.textContent()).toMatch("0");
```
Comment (pdehaan) [inline src/e2e/specs/dashboard.spec.ts:350]: nit: I think [`.toMatch()`](https://playwright.dev/docs/api/class-genericassertions#generic-assertions-to-match) is mostly for regexp comparisons. You could also use `.toBe()` or `.toEqual()`.

File: src/e2e/specs/purchase.spec.ts [hunk 14]
```diff
@@ -5,61 +5,39 @@
...
+    await landingPage.goToSignIn();
+
+    // Fill out sign up form
+    const randomEmail = `_${Date.now()}@restmail.net`;
+    await authPage.signUp(randomEmail, page);
+
+    // wait for welcome page
+    await page.waitForURL("**/user/welcome");
+    await welcomePage.goThroughFirstScan();
+    expect(page.url()).toContain("/user/dashboard");
```
Comment (pdehaan) [inline src/e2e/specs/purchase.spec.ts:39]: could also use [`.toHaveURL()`](https://playwright.dev/docs/api/class-pageassertions#page-assertions-to-have-url) with a urlOrRegExp. 🤷

**Review by pdehaan — APPROVED** [review]
R+ with a few optional comments.

**Pull request merged**